File: FrostScript/FrostScript.fs
[<AutoOpen>]
module FrostScript.FrostScript
    open FrostScript.Domain.NativeFunctions
    open FrostScript.Domain
    open FrostScript.Features

    let execute =
        let features = [
            literal
            frostlist
        ]

        let lex (script : string) = 
            let rec getTokens (ctx : LexContext) =
                if ctx.Characters = [] then
                    Ok ctx
                else
                    let whiteSpace = fun ctx ->
                        match ctx.Characters with
                        | [] -> Some (Ok ctx)
                        | char :: tail ->
                            match char with
                            | ' ' | '\t' -> Ok { ctx with Characters = tail; Position = { ctx.Position with Character = ctx.Position.Character + 1 } } |> Some
                            | _ -> None
                        
                    features
                    |> List.map (fun x -> x.Lexer)
                    |> List.append [whiteSpace]
                    |> choose ctx
                    |> Result.bind getTokens

            { Characters = script.ToCharArray () |> Array.toList; Position = { Character = 0; Line = 0; }; Tokens = [] }
            |> getTokens
            |> Result.map (fun x -> x.Tokens)

        let parse tokens =
            let rec getNode (ctx : ParseContext) =
                if ctx.Tokens = [] then
                    Ok ctx
                else
                    features
                    |> List.map (fun x -> x.Parser getNode)
                    |> choose ctx
          
            { Tokens = tokens; Node = { Token = tokens.Head; Type = StatementNode } }
            |> getNode
            |> Result.map (fun ctx -> ctx.Node)

        let rec validate node = 
            features
            |> List.map (fun x -> x.Validator validate)
            |> choose  { Node = node; Ids = { Values = [] } } 

        let rec interpret expression = 
            features
            |> List.map (fun x -> x.Interpreter interpret)
            |> choose { Expression = expression; Ids = { Values = [] } } 

        lex >> 
        apply (Ok splitTokens) >> 
        bindTraverse parse >>
        bindTraverse validate >>
        Result.map (fun validationOutput -> validationOutput |> List.map (fun x -> x.Expression)) >>
        bindTraverse interpret >> 
        Result.map (fun interpretOutput -> (interpretOutput |> List.last).Value) >>
        Result.mapError (fun errors -> 
            errors 
            |> List.map (fun (token, error) -> $"[Line:{token.Position.Line} Character:{token.Position.Character}] {error}")
            |> String.concat System.Environment.NewLine
        )

File: FrostScript.CSharp/FrostScript.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FrostScript.Expressions;
using FrostScript.NativeFunctions;
using Frostware.Result;
using static FrostScript.Lexer;
using static FrostScript.Parser;
using static FrostScript.Validator;
using static FrostScript.Interpreter;

namespace FrostScript
{
    public static class FrostScript
    {
        public static readonly IReadOnlyDictionary<string, IExpression> NativeFunctions = new Dictionary<string, IExpression>()
        {
            ["print"] = new PrintFunction(),
        };

        private static readonly Func<string, Result> execute = 
            lex
            .Fish(parse)
            .Fish(validate(new(NativeFunctions)))
            .Fish(interpret(new(NativeFunctions)));


        public static void ExecuteString(string frostScript)
        {
            execute(frostScript);
        }

        public static T ExecuteString<T>(string frostScript)
        {
            return (T)(execute(frostScript) as Pass<object>).Value;
        }

        public static void ExecuteFile(string path)
        {
            execute(File.ReadAllText(path));
        }

        public static T ExecuteFile<T>(string path)
        {
            return (T)(execute(File.ReadAllText(path)) as Pass<object>).Value;
        }
    }
}
